### PR TITLE
Ashank's getBatchById API

### PR DESCRIPTION
**Pull request by tavisca-adsouza**
I have added unit tests, all of which passed. There may still be some work with dealing with exceptions and errors.

### CODE REVIEW

File: src/Tavisca.TMS.Core/Services/BatchService.cs [hunk 1]
```diff
@@ -112,5 +112,19 @@ public ApiResponse CreateBatch(Batch batch, string trainingId)
                 return null;
             }
         }
+
+        public ApiResponse GetBatchById(string id)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Services/BatchService.cs]: All the methods in api must be asynchronous

File: src/Tavisca.TMS.Core/Services/BatchService.cs [hunk 2]
```diff
@@ -112,5 +112,19 @@ public ApiResponse CreateBatch(Batch batch, string trainingId)
...
+        {
+            
+            BatchResponse<BatchDetails> batchDetails = _batchRepository.GetBatchById(id);
+            if (batchDetails == null || batchDetails.status == Status.failure)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Services/BatchService.cs]: batchDetails.status == Status.failure consider case sensitivity also while doing comparison

File: src/Tavisca.TMS.Persistence/Dal/BatchDal.cs [hunk 3]
```diff
@@ -136,5 +135,128 @@ private BatchDto ConvertToList(MySqlDataReader reader)
...
+        public DalResponse<BatchDto> GetBatchById(string id)
+        {
+            DalResponse<BatchDto> dalResponse = new DalResponse<BatchDto>();
+            using (var conn = mySqlConnectionProvider.GetConnection())
+            {
+                try
+                {
+                    conn.Open();
+                    var query = "sp_GetBatchById";
+                    using (MySqlCommand cmd = new MySqlCommand(query, conn))
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/Dal/BatchDal.cs]: Remove using from everything except the connection.

File: src/Tavisca.TMS.Persistence/Dal/BatchDal.cs [hunk 4]
```diff
@@ -136,5 +135,128 @@ private BatchDto ConvertToList(MySqlDataReader reader)
...
+                                return dalResponse;
+                            }
+                        }
+                        
+                    }
+
+                }
+                catch (RowNotInTableException ex)
+                {
+                    dalResponse.data = null;
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/Dal/BatchDal.cs]: dalResponse.data = null will be the default behaviour.

File: src/Tavisca.TMS.Persistence/Dal/BatchDal.cs [hunk 5]
```diff
@@ -136,5 +135,128 @@ private BatchDto ConvertToList(MySqlDataReader reader)
...
+                        }
+                        
+                    }
+
+                }
+                catch (RowNotInTableException ex)
+                {
+                    return new List<string>(); 
+                }
+                catch (MySqlException mySqlex)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/Dal/BatchDal.cs]: what is a purpose of specific exception handling if you are handling it in same way.

File: src/Tavisca.TMS.Persistence/Repositories/BatchRepository.cs [hunk 6]
```diff
@@ -55,5 +55,27 @@ public List<BatchDetails> GetByTrainingID(string trainingId)
...
+                BatchDetails batchDetails = _translator.Translate(dalResponse.data);
+                batchResponse.data = batchDetails;
+                batchResponse.data.EmployeeIds = _dal.GetEmployeesIdOfBatch(id);
+            }
+            else
+            {
+                batchResponse.ErrorMessage = dalResponse.ErrorMessage;
+                batchResponse.ErrorCode = dalResponse.ErrorCode;
+                batchResponse.data = null;
+                batchResponse.status = Status.failure;
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/Repositories/BatchRepository.cs]: Remove status from here. Host should decide the status based on the exceptions, errors and response provided by the underlying layer.

File: src/Travisca.TMS.Contracts/Interfaces/Dals/IBatchDal.cs [hunk 7]
```diff
@@ -2,12 +2,19 @@
...
+
 namespace Tavisca.TMS.Contracts
 {
     public interface IBatchDal
     {
         List<BatchDto> FetchByTrainingID(string trainingId);
         string CreateBatch(Batch batch,string trainingId);
 
+        //BatchDto GetBatchById(string id);
+        List<string> GetEmployeesIdOfBatch(string id);
```
Comment (ckulkarni007) [inline src/Travisca.TMS.Contracts/Interfaces/Dals/IBatchDal.cs]: Make everything asynchronous.

Comment (ckulkarni007) [inline src/Travisca.TMS.Contracts/Interfaces/Dals/IBatchDal.cs]: id to batchId

File: src/Tavisca.TMS.Core/Services/BatchService.cs (hunk 1 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Core/Services/BatchService.cs]: Made methods asynchronous.

File: src/Tavisca.TMS.Core/Services/BatchService.cs (hunk 2 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Core/Services/BatchService.cs]: Done.

File: src/Tavisca.TMS.Persistence/Dal/BatchDal.cs (hunk 3 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Persistence/Dal/BatchDal.cs]: Done this and added await Task for dal layer.

**Pull request merged**